activerecord/lib/active_record/connection_handling.rb

```diff
@@ -135,9 +135,11 @@ def connects_to(database: {}, shards: {})
     #
     # The database kwarg is deprecated and will be removed in 6.2.0 without replacement.
     def connected_to(database: nil, role: nil, shard: nil, prevent_writes: false, &blk)
-      raise NotImplementedError, "`connected_to` can only be called on ActiveRecord::Base" unless self == Base
-
-      if database
+      if shard && role && (!abstract_class? || self != Base)
```

Comment:
`!abstract_class? || self != Base` is always true.

```suggestion
      if shard && role
```